text/0000-int-guidelines.md

```diff
@@ -0,0 +1,164 @@
+assertions affect the traditional tradeoffs about unsigned types, which are described
+under the detailed design below.
+
+It also draws inspiration from the [Google C++ Style Guide](http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Integer_Types)
+
+This RFC attempts to balance several concerns:
+
+* Rust developers should have an easy, go-to heuristic for deciding what
+  integer sizes to use when they need to use a number type.
+* 32-bit integers are considerably faster than 64-bit integers in some
+  situations, so using 64-bit integers for tiny numbers, especially in
+  hot code, can result in unnecessarily slower programs.
+* Reflexive use of 32-bit integers too-often results in overflows when the
+  numbers aren't expected to be "laughably smaller" than the maximum
+  32-bit number.
+* Occasional use of `usize` when building a brand new data structure may
+  be appropriate, but use of `usize` in general can introduce portability
+  hazards when the use-case is not proportional to the amount of
+  addressable memory.
+* Using of unsigned integers is traditionally thought to be error-prone, and
+  style guides often suggest avoiding them. That said, Rust's unsigned integers
+  have built-in underflow assertions, which changes the analysis.
+
+# Detailed design
+
+When deciding what integer type to use for a given situation, you can make the decision
+to a first approximation through this heuristic:
+
+* If you expect all uses of this API to use numbers "laughably smaller" then 32-bits,
+  use 32-bit numbers. Otherwise, prefer 64-bit numbers.
+* If you are sure that the number will never be less than 0, use unsigned integers.
+  Otherwise, use signed numbers.
+* If you are building a new data structure, and your number refers to the size of memory,
+  you may want to use `usize`, which is described below.
+
+> Note: Rust does not yet have BigInts in `std`, which might, in theory, be a better
+> go-to big integer type. These guidelines will probably be revised in the future
+> once that changes.
+
+## Signed vs. Unsigned Integers
+
+Traditionally, style guides for low-level languages have [warned against the use
+of unsigned integers](http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Integer_Types) because
+they can introduce bugs when comparing unsigned value with 0
+(`for (unsigned int i = foo.len() - 1; i >= 0; --i) ...`). The Google Style Guide
+argues that instead of unsigned integers, programmers should use signed
+integers and assertions that the value does not go below 0.
+
+In Rust, unsigned integers have underflow checking assertions built-into
+the type (assuming that RFC XXX is accepted), so using a `u32` is equivalent
+to the advice in Google style guide (with a larger maximum value).
+
+## The "size" Types
+
+The `isize` and `usize` types represent numbers that are proportional to
+the size of available memory. Most normal uses of `isize` and `usize`
```

Comment:
Yes. Also I think it's good to bring up the fact that these types scale both *up and down* with the address space. ("Address space" is more accurate than "available memory.") They should beware of using `isize` or `usize` where it might not suffice on 32-bit or 16-bit machines.